File: 166__Fraction_to_Recurring_Decimal.py
class Solution(object):
    def fractionToDecimal(self, numerator, denominator):
        """
        :type numerator: int
        :type denominator: int
        :rtype: str
        """
        if numerator == 0:
            return "0"

        neg = (numerator < 0) ^ (denominator < 0)
        w, r = divmod(abs(numerator), abs(denominator))
        s = str(w)
        if neg:
            s = "-" + s
        if r:
            s += '.'
        i = len(s)
        nums = {r: i}
        while r:
            w, r = divmod(r * 10, abs(denominator))
            s += str(w)
            if r in nums:
                s = s[:nums[r]] + "(" + s[nums[r]:] + ")"
                return s
            i += 1
            nums[r] = i
        return s
        

File: Java/166__Fraction_to_Recurring_Decimal.java
class Solution {
    public String fractionToDecimal(int numerator, int denominator) {
        if (numerator == 0) {
            return "0";
        }
        StringBuilder result = new StringBuilder();
        if (numerator < 0 ^ denominator < 0) {
            result.append("-");
        }
        long dividend = Math.abs(Long.valueOf(numerator));
        long divisor = Math.abs(Long.valueOf(denominator));
        result.append(String.valueOf(dividend / divisor));
        long remainder = dividend % divisor;
        if (remainder == 0) {
            return result.toString();
        }
        result.append(".");
        Map<Long, Integer> map = new HashMap<>();
        while (remainder != 0) {
            if (map.containsKey(remainder)) {
                result.insert(map.get(remainder), "(");
                result.append(")");
                break;
            }
            map.put(remainder, result.length());
            remainder *= 10;
            result.append(String.valueOf(remainder / divisor));
            remainder %= divisor;
        }
        return result.toString();
        }

}
